bigtable/examples/run_examples_utils.sh

```diff
@@ -19,6 +19,47 @@ readonly CBT_CMD="${CBT:-${GOPATH}/bin/cbt}"
 readonly CBT_EMULATOR_CMD="${CBT_EMULATOR:-${GOPATH}/bin/emulator}"
 readonly CBT_INSTANCE_ADMIN_EMULATOR_CMD="../tests/instance_admin_emulator"
 
+function cleanup_instance {
+  local project=$1
+  local instance=$2
+  shift 2
+
+  if [ -z "${project}" -o -z "${instance}" ]; then
+    return
+  fi
+
+  echo
+  echo "Cleaning up test instance projects/${project}/instances/${instance}"
+  local setenv="env"
+  if [ -n "${BIGTABLE_INSTANCE_ADMIN_EMULATOR_HOST:-}" ]; then
+    setenv="env BIGTABLE_EMULATOR_HOST=${BIGTABLE_INSTANCE_ADMIN_EMULATOR_HOST}"
+  fi
+  ${setenv} ../examples/bigtable_samples_instance_admin delete-instance "${project}" "${instance}"
+}
+
+function exit_handler {
+  local project=$1
+  local instance=$2
+  shift 2
+
+  cleanup_instance "${project}" "${instance}"
+  if [ -n "${BIGTABLE_INSTANCE_ADMIN_EMULATOR_HOST:-}" ]; then
+    # If the test is running against the emulator there is no need to cleanup
+    # the instance, just kill the emulators.
+    kill_emulators
+  fi
+}
+
+function reset_trap {
+  if [ -n "${BIGTABLE_INSTANCE_ADMIN_EMULATOR_HOST:-}" ]; then
```

Comment:
This looks like it's duplicating logic that could be elsewhere (see earlier comment).

Can we just have a single unified cleanup function, that chooses how to cleanup based on the same env var? If it's set, it's clearly using emulator, so that's what needed to be cleaned up, if it's not set, it's using the real service, so the real instance needs to be cleaned up.

Am I missing anything?

Reply:
I want this script to more or less follow the steps we would give a user in a tutorial (eventually it should become a test for "the tutorial works"):

- Create an instance.
- Run some examples (which examples depends on the tutorial).
- Delete the instance.

Therefore, I want to cleanup the instance at the end of each group of examples (`instance_admin`, `table_admin` and `data`).  If I have the unified function that kills the emulator then the next group of examples has nothing to run against.

I could instead do something like:

- Create an instance
- Run the instance admin examples using that instance, probably with a second instance to demonstrate delete-instance.
- Run the table admin examples.
- Run the data examples.
- Delete the instance (or shutdown the emulators).

But that is fairly different from what the users would experience.

The code flow gets complicated by trying to handle the case where the program aborts and cleaning up the instance (or the emulators) in that case.  Open to suggestions here because the flow is admittedly complicated.